Observer/SpeakFasterObserver/AudioInput.cs

```
@@ -0,0 +1,100 @@
+﻿using FlacBox;
+using NAudio.Wave;
+using System;
+using System.IO;
+
+namespace SpeakFasterObserver
+{
+    class AudioInput
+    {
+        private static int AUDIO_NUM_CHANNELS = 1;
+        private static int AUDIO_BITS_PER_SAMPLE = 16;
+        private static int AUDIO_SAMPLE_RATE_HZ = 16000;
+        // NOTE: Due to limited buffer size, using a write timer of with a
+        // period of 120 s or greater will cause data loss, which will throw an
+        // exception.
+        private static int MAX_BUFFER_LENGTH = AUDIO_SAMPLE_RATE_HZ * 120;
+
+        private WaveIn waveIn = null;
+        private bool isRecording = false;  // TODO(cais): Thread safety?
+        private int[] buffer = null;
+        private int bufferPointer = 0;
+        private static readonly object lockObj = new object();
+
+        public AudioInput() {}
+
+        public void StartRecordingFromMicrophone()
+        {
+            if (isRecording)
+            {
+                return;
+            }
+            waveIn = new WaveIn();
+            waveIn.WaveFormat = new WaveFormat(AUDIO_SAMPLE_RATE_HZ, AUDIO_NUM_CHANNELS);
+            if (waveIn.WaveFormat.BitsPerSample != AUDIO_BITS_PER_SAMPLE)
+            {
+                throw new NotSupportedException(
```

Unhandled exception.  Bad user experience if triggered.  Would prefer a messagebox + termination or a silent failure or a Debug.Assert to an unhandled exception user experience.

Note that we don't have an exception reporting service hooked up to this app yet, so an unhandled exception just spews to the user and we don't get notification.  If we are going to allow unhandled exceptions, we should add an app-wide unhandled exception handler that turns into a copyable messagebox with an email address to send to.  Hmm, we have an example of this somewhere that we should be able to find.  Let's file a bug on this too.

---

Fair point. I filed #64 for adding an app-wide exception handling mechanism and added a TODO item here. For now I'll just leave this exception as is. I think the 16-bit audio format is fairly common and the likelihood of finding it not support on a Surface device is low.